zio-http/src/main/scala/zhttp/http/Http.scala

```diff
@@ -869,7 +863,57 @@ object Http {
    * Creates an Http app from a resource path
    */
   def fromResource(path: String): HttpApp[Any, Throwable] =
-    Http.getResource(path).flatMap(url => Http.fromFile(new File(url.getPath)))
+    Http.getResource(path).flatMap(url => Http.fromResourceWithURL(url))
+
+  private[zhttp] def fromResourceWithURL(url: java.net.URL): HttpApp[Any, Throwable] = {
+    url.getProtocol match {
+      case "file" =>
+        Http.fromFile(new File(url.getPath))
+      case "jar"  =>
+        val path         = new java.net.URI(url.getPath).getPath // remove "file:" prefix and normalize whitespace
+        val bangIndex    = path.indexOf('!')
+        val filePath     = path.substring(0, bangIndex)
+        val resourcePath = path.substring(bangIndex + 2)
+        val appZIO       =
+          ZIO
+            .attemptBlockingIO(new java.util.zip.ZipFile(filePath))
+            .asSomeError
+            .flatMap { jar =>
+              val closeJar = ZIO.attemptBlocking(jar.close()).ignoreLogged
+              (for {
+                entry <- ZIO.attemptBlocking(Option(jar.getEntry(resourcePath))).some
+                _     <- ZIO.when(entry.isDirectory)(ZIO.fail(None))
+                contentLength = entry.getSize
+                inStream <- ZIO.attemptBlockingIO(jar.getInputStream(entry)).asSomeError
+                mediaType = determineMediaType(resourcePath)
+                app       =
+                  Http
+                    .fromStream(
+                      ZStream
+                        .fromInputStream(inStream)
+                        .ensuring(closeJar),
+                    )
+                    .withContentLength(contentLength)
+              } yield mediaType.fold(app)(t => app.withMediaType(t))).onError(_ => closeJar)
+            }
+
+        Http.fromZIO(appZIO).flatten.catchAll {
+          case Some(e) => Http.fail(e)
+          case None    => Http.empty
+        }
+      case proto  =>
+        Http.fail(new IllegalArgumentException(s"Unsupported protocol: $proto"))
+    }
+  }
+
+  private def determineMediaType(filePath: String): Option[MediaType] = {
+    // Extract file extension
+    val ext = filePath.lastIndexOf(".") match {
+      case -1 => None
+      case i  => Some(filePath.substring(i + 1))
+    }
+    ext.flatMap(MediaType.forFileExtension)
```

Comment:
instead of doing a flatmap we can move this operation above. That will save us a lambda and an additional Option.

Reply:
Thanks, fixed.